client/src/_session_layer.rs

```diff
@@ -0,0 +1,1536 @@
+                            .ok();
+                    }
+                    .boxed_local()
+                }
+                _ => {
+                    log::debug!("Unhandled control packet: {kind:?}");
+                    return None;
+                }
+            };
+            return Some(fut);
+        }
+        None
+    }
+
+    fn on_request(
+        self,
+        session_id: Vec<u8>,
+        request: proto::Request,
+        from: SocketAddr,
+    ) -> Option<LocalBoxFuture<'static, ()>> {
+        log::trace!("Received request packet from {from}: {request}");
+
+        let (request_id, kind) = match request {
+            proto::Request {
+                request_id,
+                kind: Some(kind),
+            } => (request_id, kind),
+            _ => return None,
+        };
+
+        if self.is_request_duplicate(&session_id, request_id) {
+            return None;
+        }
+        self.record_duplicate(session_id.clone(), request_id);
+
+        let fut = match kind {
+            proto::request::Kind::Ping(request) => {
+                async move { self.on_ping(session_id, request_id, from, request).await }
+                    .boxed_local()
+            }
+            proto::request::Kind::Session(request) => async move {
+                self.dispatch_session(session_id, request_id, from, request)
+                    .await
+                    .map_err(|e| log::warn!("Handling `Session` request: {e}"))
+                    .ok();
+            }
+            .boxed_local(),
+            _ => return None,
+        };
+
+        Some(fut)
+    }
+
+    fn on_forward(
+        self,
+        forward: Forward,
+        from: SocketAddr,
+        session: Option<Arc<DirectSession>>,
+    ) -> Option<LocalBoxFuture<'static, ()>> {
+        let reliable = forward.is_reliable();
+        let _encrypted = forward.is_encrypted();
+        let slot = forward.slot;
+        let channel = self.ingress_channel.clone();
+
+        let myself = self;
+        let fut = async move {
+            log::trace!(
+                "[{}] received forward packet ({} B) via {}",
+                myself.config.node_id,
+                forward.payload.len(),
+                from
+            );
+
+            let session = match session {
+                None => {
+                    // In this case we can't establish session, because we don't have
+                    // neither NodeId nor SlotId.
+                    log::debug!(
+                        "Forward packet from unknown address: {from}. Can't resolve. Sending Disconnected message.",
+                    );
+                    return myself.send_disconnect(SessionId::from(forward.session_id), from).await;
+                },
+                Some(session) => session,
+            };
+
+            let node = if slot == FORWARD_SLOT_ID {
+                // Direct message from other Node.
+                session.owner.default_id
+            } else {
```

Comment:
```suggestion
            let other_node = if direct_message(slot) {
                session.owner.default_id
            } else {
```

Reply:
Done in https://github.com/golemfactory/ya-relay/pull/250/commits/55427fc6f42ba55fe6867e4c7363207f5b1ae4b9